File: app/src/main/java/com/github/overpass/gather/auth/register/signup/SignUpFragment.java
package com.github.overpass.gather.auth.register.signup;

import android.os.Handler;

import androidx.lifecycle.ViewModelProviders;

import com.github.overpass.gather.R;
import com.github.overpass.gather.auth.register.RegistrationFragment;
import com.github.overpass.gather.dialog.ProgressDialogFragment;
import com.google.android.material.textfield.TextInputEditText;

import butterknife.BindView;
import butterknife.OnClick;

import static com.github.overpass.gather.UIUtil.snackbar;
import static com.github.overpass.gather.UIUtil.textOf;

public class SignUpFragment extends RegistrationFragment<SignUpViewModel> {

    @BindView(R.id.tietEmail)
    TextInputEditText tietEmail;
    @BindView(R.id.tietPassword)
    TextInputEditText tietPassword;

    @Override
    protected SignUpViewModel createViewModel() {
        return ViewModelProviders.of(this).get(SignUpViewModel.class);
    }

    @Override
    protected int getLayoutRes() {
        return R.layout.fragment_sign_up;
    }

    public void handleSignUp(SignUpStatus status) {
        switch (status.tag()) {
            case SignUpStatus.ERROR:
                ProgressDialogFragment.hide(getFragmentManager());
                String message = status.as(SignUpStatus.Error.class)
                        .getThrowable()
                        .getLocalizedMessage();
                snackbar(tietEmail, message);
                break;
            case SignUpStatus.SUCCESS:
                ProgressDialogFragment.hide(getFragmentManager());
                snackbar(tietEmail, "Success");
                new Handler().postDelayed(() -> {
                    viewModel.moveToNextStep(registrationController);
                }, 200);
                break;
            case SignUpStatus.PROGRESS:
                ProgressDialogFragment.show(getFragmentManager());
                break;
            case SignUpStatus.INVALID_EMAIL:
                ProgressDialogFragment.hide(getFragmentManager());
                snackbar(tietEmail, status.as(SignUpStatus.InvalidEmail.class).getMessage());
                break;
            case SignUpStatus.INVALID_PASSWORD:
                ProgressDialogFragment.hide(getFragmentManager());
                snackbar(tietEmail, status.as(SignUpStatus.InvalidPassword.class).getMessage());
                break;
        }
    }

    @OnClick(R.id.tvSignUp)
    public void onSignUpClicked() {
        viewModel.signUp2(textOf(tietEmail), textOf(tietPassword))
                .observe(getViewLifecycleOwner(), this::handleSignUp);
    }

    public static SignUpFragment newInstance() {
        return new SignUpFragment();
    }
}

File: app/src/main/java/com/github/overpass/gather/ui/auth/register/signup/SignUpFragment.kt
package com.github.overpass.gather.ui.auth.register.signup

import android.os.Bundle
import android.os.Handler
import android.view.View
import com.github.overpass.gather.App.Companion.componentManager
import com.github.overpass.gather.R
import com.github.overpass.gather.commons.android.lifecycle.on
import com.github.overpass.gather.commons.android.snackbar
import com.github.overpass.gather.commons.android.text
import com.github.overpass.gather.ui.auth.register.RegistrationFragment
import com.github.overpass.gather.ui.dialog.progress.indeterminate.ProgressDialogFragment
import kotlinx.android.synthetic.main.fragment_sign_up.*
import kotlinx.coroutines.ExperimentalCoroutinesApi

class SignUpFragment : RegistrationFragment<SignUpViewModel>() {

    override fun getLayoutRes(): Int = R.layout.fragment_sign_up

    override fun createViewModel(): SignUpViewModel {
        return viewModelProvider.get(SignUpViewModel::class.java)
    }

    override fun inject() {
        componentManager.getSignUpComponent()
                .inject(this)
    }

    @ExperimentalCoroutinesApi
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        tvSignUp.setOnClickListener {
            viewModel.signUp(tietEmail.text(), tietPassword.text())
        }
    }

    override fun onBind() {
        super.onBind()
        on(viewModel.signUpError()) {
            handleError(it)
        }
        on(viewModel.signUpSuccess()) {
            handleSuccess()
        }
        on(viewModel.signUpProgress()) {
            handleProgress()
        }
        on(viewModel.invalidEmail()) {
            handleInvalidEmail(it)
        }
        on(viewModel.invalidPassword()) {
            handleInvalidPassword(it)
        }
    }

    private fun handleError(message: String) {
        tietEmail.snackbar(message)
    }

    private fun handleSuccess() {
        ProgressDialogFragment.hide(fragmentManager)
        tietEmail.snackbar("Success")
        viewModel.setSignUpInProgress()
        Handler().postDelayed({
            registrationController!!.moveToNextStep()
        }, 200)
    }

    private fun handleProgress() {
        ProgressDialogFragment.show(fragmentManager)
    }

    private fun handleInvalidEmail(message: String) {
        ProgressDialogFragment.hide(fragmentManager)
        tietEmail.error = message
    }

    private fun handleInvalidPassword(message: String) {
        ProgressDialogFragment.hide(fragmentManager)
        tietPassword.error = message
    }

    companion object {

        @JvmStatic
        fun newInstance(): SignUpFragment {
            return SignUpFragment()
        }
    }
}
